### PR TITLE
References

### PR DESCRIPTION
**Pull request by virajkaram**
This PR refactors the WFCAM references, mainly to separate the query and the stacking parts. 
Also hopefully solves errors like COMPID is missing.
Adds caching for winter references. 
Some minor feature improvements to reference-making, photcal, swarp and zogy.
Stacking offline images WFCAM will follow in a separate PR shortly.

### CODE REVIEW

File: mirar/io.py [hunk 1]
```diff
@@ -86,14 +86,14 @@ def save_mef_to_path(data_list, header_list, primary_header, path):
     hdulist.writeto(path, overwrite=True)
 
 
-def open_fits(path: str | Path) -> tuple[np.ndarray, fits.Header]:
+def open_fits(path: str | Path, **fits_open_kwargs) -> tuple[np.ndarray, fits.Header]:
```
Comment (robertdstein) [inline mirar/io.py:89]: What are the `fits_open_kwargs` that need to be passed?

File: mirar/io.py [hunk 2]
```diff
@@ -108,9 +108,33 @@ def open_fits(path: str | Path) -> tuple[np.ndarray, fits.Header]:
...
+    path: str | Path,
+):
+    """
+    Save an Image to path
+
+    :param image: Image to save
+    :param path: path
+    :return: None
+    """
+    path = str(path)
```
Comment (robertdstein) [inline mirar/io.py:122]: Why do you turn the path into a string first?

File: mirar/io.py [hunk 3]
```diff
@@ -108,9 +108,33 @@ def open_fits(path: str | Path) -> tuple[np.ndarray, fits.Header]:
...
+        header[LATEST_SAVE_KEY] = path
+    logger.debug(f"Saving to {path}")
+    save_to_path(data, header, path)
+
+
 def open_raw_image(
     path: str | Path,
-    open_f: Callable[[str | Path], tuple[np.ndarray, fits.Header]] = open_fits,
+    open_f: Callable[
+        [str | Path, Optional[dict]], tuple[np.ndarray, fits.Header]
```
Comment (robertdstein) [inline mirar/io.py:135]: Is this righ? open_f can work on a dictionary?

File: mirar/pipelines/winter/generator.py [hunk 4]
```diff
@@ -29,20 +29,42 @@
...
+
+
+def winter_wfau_component_image_stacker(**kwargs) -> Swarp:
+    """
+    Generates a resampler for reference images
+
+    :param kwargs: kwargs
+    :return: Swarp processor
+    """
+    logger.debug(kwargs)
```
Comment (robertdstein) [inline mirar/pipelines/winter/generator.py:58]: this is a bad logger message. If you want it please add some text explaining what you are printing e.g `wfau kwargs: {kwargs}`

File: mirar/pipelines/winter/generator.py [hunk 5]
```diff
@@ -262,9 +248,6 @@ def winter_refbuild_reference_generator(image: Image):
     components_image_dir = get_output_dir(
         dir_root="components", sub_dir="winter/references"
     )
-    stack_image_dir = get_output_dir(
-        dir_root="stacked_ref", sub_dir="winter/references"
-    )
     if not components_image_dir.exists():
         components_image_dir.mkdir(parents=True)
```
Comment (robertdstein) [inline mirar/pipelines/winter/generator.py:252]: make this a one-liner to avoid race conditions

File: mirar/processors/astromatic/swarp/swarp.py [hunk 6]
```diff
@@ -358,8 +366,10 @@ def _apply_to_images(
             )
 
             if temp_output_image_path.exists():
-                temp_output_image_path.rename(output_image_path)
-                temp_output_image_weight_path.rename(output_image_weight_path)
+                shutil.copy(temp_output_image_path, output_image_path)
```
Comment (robertdstein) [inline mirar/processors/astromatic/swarp/swarp.py:369]: swarp should not save images, that should only happen in imagesaver

File: mirar/processors/zogy/zogy.py [hunk 7]
```diff
@@ -49,6 +49,7 @@ class ZOGYError(ProcessorError):
 
 def default_wirc_catalog_purifier(sci_catalog: Table, ref_catalog: Table):
     """
+    TODO: This should be in wirc?
```
Comment (robertdstein) [inline mirar/processors/zogy/zogy.py:52]: Please move to wirc

File: mirar/processors/zogy/zogy.py [hunk 8]
```diff
@@ -73,6 +74,9 @@ def default_wirc_catalog_purifier(sci_catalog: Table, ref_catalog: Table):
 
 
 def default_summer_catalog_purifier(sci_catalog: Table, ref_catalog: Table):
+    """
+    TODO: Catalog purifier for the summer data. This should probably be in summer.
```
Comment (robertdstein) [inline mirar/processors/zogy/zogy.py:78]: move to summer

File: mirar/processors/zogy/zogy.py [hunk 9]
```diff
@@ -94,6 +98,9 @@ def default_summer_catalog_purifier(sci_catalog: Table, ref_catalog: Table):
 
 
 def default_sedmv2_catalog_purifier(sci_catalog, ref_catalog):
+    """
+    TODO: This should be in sedmv2?
```
Comment (robertdstein) [inline mirar/processors/zogy/zogy.py:102]: move to sedmv2

File: mirar/references/base_reference_generator.py [hunk 10]
```diff
@@ -173,3 +195,139 @@ def get_output_path(output_dir: str | Path, base_name: str) -> Path:
...
+        Get component reference images that will be stacked together.
+        """
+        raise NotImplementedError
+
+    def _get_reference(self, image: Image) -> (fits.PrimaryHDU, fits.PrimaryHDU):
+        """
+        Get loaded ref image for image
+
+        :param image: image
+        :return: ref image
```
Comment (robertdstein) [inline mirar/references/base_reference_generator.py:278]: you do not return a ref image, you return two fits.HDUs

File: mirar/references/wfcam/utils.py [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+    ra_list, dec_list = wcs.all_pix2world(xcrd_list, ycrd_list, 1)
+    ra_list[ra_list < 0] = ra_list[ra_list < 0] + 360
+    return ra_list, dec_list
+
+
+def find_ukirt_surveys(ra: float, dec: float, band: str) -> list[MOCSurvey]:
+    """
+    Find which UKIRT survey does the given RA/Dec belong to
+    Args:
+        ra:
```
Comment (robertdstein) [inline mirar/references/wfcam/utils.py:76]: Add docstring

(style :param / :return)

File: mirar/references/wfcam/utils.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+    band_surveys = np.array([x for x in surveys if x.filter_name == band])
+    in_survey_footprint = [x.contains(ra, dec)[0] for x in band_surveys]
+    return band_surveys[in_survey_footprint]
+
+
+def combine_headers(primary_header, header_to_append):
+    """
+    Function to append a header to another
+    Args:
+        primary_header:
```
Comment (robertdstein) [inline mirar/references/wfcam/utils.py:93]: Add docstring

File: mirar/references/wfcam/utils.py [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+    return primary_header
+
+
+def make_wfcam_image_from_hdulist(
+    ukirt_hdulist: [fits.hdu.image.PrimaryHDU, fits.hdu.image.ImageHDU],
+    url: str,
+) -> Image:
+    """
+    Function to convert a ukirt image with two headers to a single header image
+    Args:
```
Comment (robertdstein) [inline mirar/references/wfcam/utils.py:119]: add docstring

File: mirar/references/wfcam/wfcam_query.py [hunk 14]
```diff
@@ -0,0 +1,626 @@
...
+                COMPID_KEY.lower(),
+            ]
+            for key in required_query_db_keys:
+                if key not in self.query_db_table.sql_model.__table__.columns:
+                    raise ValueError(f"{key} must be present in the query_db_table")
+
+    def get_query_class(self) -> BaseWFAUClass:
+        """
+        Get the class that will be used to query the WFAU database, e.g. VSAClass or
+        UKIDSSClass
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:228]: Add :return

File: mirar/references/wfcam/wfcam_query.py [hunk 15]
```diff
@@ -0,0 +1,626 @@
...
+        """
+        raise NotImplementedError
+
+    def get_surveys(self, ra: float, dec: float) -> list[MOCSurvey]:
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:232]: Add docstring

File: mirar/references/wfcam/wfcam_query.py [hunk 16]
```diff
@@ -0,0 +1,626 @@
...
+        """
+        Get the surveys that are available at the given coordinates
+        """
+        raise NotImplementedError
+
+    def get_query_crds(
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:238]: add :return

File: mirar/references/wfcam/wfcam_query.py [hunk 17]
```diff
@@ -0,0 +1,626 @@
...
+        self,
+        image: Image,
+    ) -> ImageBatch:
+        """
+        Run the query to the WFAU database.
+        wfau_query: WFAU query object.
+        query_crds: Coordinates to query.
+        wfau_survey_names: WFAU names of the survey databases to query.
+        filter_name: Name of the filter to query.
+        savedir: Directory to save the images.
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:259]: :param and add :return

File: mirar/references/wfcam/wfcam_query.py [hunk 18]
```diff
@@ -0,0 +1,626 @@
...
+            wfau_query.database = survey
+            paths_list, wfau_qra_list, wfau_qdec_list, query_exists_list = (
+                [],
+                [],
+                [],
+                [],
+            )
+            for ind, crd in enumerate(query_crds):
+                logger.debug(f"Running query {ind}/{len(query_crds)}")
+                crd = query_crds[ind]
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:303]: this line is redundant?

File: mirar/references/wfcam/wfcam_query.py [hunk 19]
```diff
@@ -0,0 +1,626 @@
...
+
+        if len(wfcam_image_paths) == 0:
+            err = "No image found at the given coordinates in the UKIRT database"
+            raise WFAURefNotFoundError(err)
+
+        wfcam_images = ImageBatch([open_raw_image(url) for url in wfcam_image_paths])
+
+        return wfcam_images
+
+    def run_query(self, image: Image):
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:389]: docstring

File: mirar/references/wfcam/wfcam_query.py [hunk 20]
```diff
@@ -0,0 +1,626 @@
...
+    image_width: Quantity = wfau_image_width,
+    image_height: Quantity = wfau_image_height,
+    use_local_database: bool = False,
+    components_table: Type[BaseDB] = None,
+    duplicate_protocol: str = "ignore",
+    q3c_bool: bool = False,
+) -> list[Path]:
+    """
+    Download the image from UKIRT server. Optionally, check if the image exists locally
+    and ingest it into a database.
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:408]: docstring with :params

File: mirar/references/wfcam/wfcam_query.py [hunk 21]
```diff
@@ -0,0 +1,626 @@
...
+
+            save_fits(wfcam_image, imagepath)
+            logger.debug(f"Saved UKIRT image to {local_imagepaths}")
+
+        imagepaths.append(imagepath)
+
+    return np.unique(imagepaths).tolist()
+
+
+def check_query_exists_locally(
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:494]: Add type hints

File: mirar/references/wfcam/wfcam_query.py [hunk 22]
```diff
@@ -0,0 +1,626 @@
...
+        compare_keys=[
+            QUERY_RA_KEY.lower(),
+            QUERY_DEC_KEY.lower(),
+            QUERY_FILT_KEY.lower(),
+        ],
+        comparators=["__eq__", "__eq__", "__eq__"],
+    )
+    logger.debug(results)
+    savepaths = []
+    # TODO - run a joint query to get the savepaths
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:521]: Is this still relevant?

File: mirar/references/wfcam/wfcam_query.py [hunk 23]
```diff
@@ -0,0 +1,626 @@
...
+    """
+    Function to get the locally existing images that overlap with the given coordinates
+    Args:
+        query_ra: ra that was queried
+        query_dec: dec that was queried
+        query_filt: filter that was queried
+        components_table: table with component image details
+
+    Returns:
+
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:550]: Add :return

File: mirar/references/wfcam/wfcam_query.py [hunk 24]
```diff
@@ -0,0 +1,626 @@
...
+        db_table:
+        multiframe_id:
+        extension_id:
+        frame_lx:
+        frame_hx:
+        frame_ly:
+        frame_hy:
+
+    Returns:
+
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:586]: Add return and :params (explain)

File: mirar/references/wfcam/wfcam_query.py [hunk 25]
```diff
@@ -0,0 +1,626 @@
...
+    return savepaths
+
+
+class UKIRTOnlineQuery(WFAUQuery):
+    """
+    Class to query the UKIRT online database at the WFAU.
+    This is a subclass of the WFAUQuery.
+    """
+
+    def get_surveys(self, ra: float, dec: float) -> list[MOCSurvey]:
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:622]: docstrings

File: mirar/references/wfcam/wfcam_query.py [hunk 26]
```diff
@@ -0,0 +1,626 @@
...
+        return find_ukirt_surveys(ra=ra, dec=dec, band=self.filter_name)
+
+    def get_query_class(self) -> BaseWFAUClass:
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_query.py:625]: docstrings

File: mirar/references/wfcam/wfcam_stack.py [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+
+
+def default_filter_wfau_images(image_batch: ImageBatch) -> ImageBatch:
+    """
+    Function to filter WFAU images
+    Args:
+        image_batch:
+
+    Returns:
+
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_stack.py:42]: returns image batch

Comment (robertdstein) [reply mirar/references/wfcam/wfcam_stack.py:42]: also, explain what the cuts are

File: mirar/references/wfcam/wfcam_stack.py [hunk 28]
```diff
@@ -0,0 +1,144 @@
...
+        )
+        self.wfcam_query = wfcam_query
+
+        self.component_image_dir = component_image_sub_dir
+        self.filter_images = filter_images
+
+    def get_component_images(self, image: Image) -> ImageBatch:
+        """
+        Get the component images for the given image
+        """
```
Comment (robertdstein) [inline mirar/references/wfcam/wfcam_stack.py:111]: docstring :param/:return

**Review by robertdstein — CHANGES_REQUESTED** [review]
The main thing is the lack of docstrings. We do documentation built with sphinx, so should follow the sphinx style: https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html

that means basically always have :param for each param, plus :return

File: mirar/io.py (hunk 1 above)
Comment (robertdstein) [reply mirar/io.py:89]: Just remove this new kwarg and put it directly in open_fits for everyone I think

File: mirar/processors/astromatic/swarp/swarp.py (hunk 6 above)
Comment (robertdstein) [reply mirar/processors/astromatic/swarp/swarp.py:376]: Ok, sounds good

**Review by robertdstein — CHANGES_REQUESTED** [review]
1 minor change then I'm happy

**Comment by virajkaram** [comment]
I wouldn't do that, it is not fits convention to have missing SIMPLE keywords, but seems to be an issue specifically with UKIRT and we can;t do anything about it..

**Review by robertdstein — APPROVED** [review]

**Pull request merged**